\documentclass[11pt]{article}

\usepackage[parfill]{parskip}

\setlength{\topmargin}{-.5in}
\setlength{\textheight}{9in}
\setlength{\oddsidemargin}{.125in}
\setlength{\textwidth}{6.25in}

\newcommand {\cmd}[1] {\begin{quote}#1\end{quote}}

\begin{document}
\title{Open Genomics Engine - Framework Documentation}
\author{Lee Baker and David Mittelman\\
Virginia Bioinformatics Institute \\
lee@leecbaker.com, david.mittelman@vt.edu}
\maketitle

\section {Introduction}
The Open Genomics Engine (OpenGE) is a open software framework for manipulating and interpreting high-throughput sequencing data. This documentation describes core functions that are integrated into the official distribution of the framework. For a background on the project visit http://www.openge.org. OpenGE is available as binaries for Linux and MacOS or as source. The source distribution contains further instructions for compiling and installing the software. Future releases will include VMware and Amazon EC2 images.
\section {Commands}
OpenGE consists of a collection of operations that can be invoked on the command line using the following syntax:

\cmd{openge \textit{command [options]}}

where \textit{command} is one of the commands listed in the following table, and \textit{options} are various flags and parameters that may be supplied depending on the command:

\begin{center}
\begin{tabular}{lp{3.5in}}
\hline
Command&Description\\ \hline
convert&Convert a SAM file to a BAM file\\
count&Count reads in a BAM file\\
dedup&Mark (or remove) duplicate reads\\
help&View detailed help for any OpenGE command\\
mergesort&Merge and sort multiple input BAM files\\
stats&Generate statistics for one or more BAM files\\
view&View a filtered subset of a BAM file\\
\end{tabular}
\end{center}

\subsection {General}
The following flags are common for all OpenGE commands. Additional flags and parameters may be available for individual commands.
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-i \textit{filename}&{-}{-}in \textit{filename}&Input filename. Input files can also be specified without this flag. Defaults to stein if no input is specified.\\
-o \textit{filename}&{-}{-}out \textit{filename}&Output filename. Defaults to stdout if omitted. \\
-v&{-}{-}verbose&Display additional information to stderr during operation (e.g. progress indicators and other informational messages). Optional.\\
-t&{-}{-}threads&Set the number of threads to be used for multithreaded operations. Optional.\\
-d&{-}{-}nothreads&Disable multithreading.\\
\end{tabular}
\end{center}

When input or output files are required, stdin or stdout may be used by simply omitting \textit{filename}. For example, the following command:

\cmd{openge mergesort}

with no input specified, will read a BAM file from stdin, operate on the file, and write the results to stdout. This enables commands to be chained together. For instance:

\cmd{openge convert in.sam $|$ openge count}

will count the number of lines in a SAM file. Console redirection can be used as well. The following two commands are equivalent in a bash shell:

\cmd{openge mergesort a.bam -o c.bam}
\cmd{openge mergesort $<$ a.bam $>$ c.bam}

\subsection {convert}
Converts (compresses) a SAM file to a BAM file.

Example:
\cmd{openge convert in.sam -o out.bam}
\subsection {count}
The count command returns the number of reads contained in one or more BAM files. No parameters (other than input files) are required.

Example:
\cmd{openge count a.bam}
\cmd{openge count b.bam c.bam d.bam}

\subsection {dedup}
The dedup command either marks duplicate reads (compatible with Picard) or removes the duplicate reads. 

\subsection {help}
The help command shows available options for any OpenGE command. 

\cmd{openge help \textit{[command]}}

For example:

\cmd{openge help mergesort}

\cmd{openge help count}

Available commands can be seen in the table at the top of this section.


\subsection {mergesort}
This command generates a single BAM file from one or more input files. The supplied input files are merged and then sorted by read position or region.

Parameters:
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-c \textit{level}&{-}{-}compression \textit{level}&Compression level- defaults to 6. Valid levels are 0-9, and correspond to zlib's deflate compression levels. \\
-r \textit{region}&{-}{-}region \textit{region}&Genomic region to use\\
-n \textit{reads}&{-}{-}n \textit{reads}&Number of reads to use per temporary file. Defaults to 500,000.\\
-C&{-}{-}compresstempfiles&Compress temporary files. By default, temporary files are not compressed.\\
\end{tabular}
\end{center}

The number of reads per temporary file can have a substantial effect on speed when processing large datasets. Increasing the number of reads per file produces larger (and fewer) temporary files at the expense of memory, but reduces the processing time.

\subsection{stats}
The stats command generates statistics for one or more input BAM files. The computed statistics are summarized below.

\subsection{view}
The view command can be used to stream all or a subset of a BAM file. Available filters are summarized below.

Parameters:
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-n&--n&Number of reads to include in the generated file\\
\end{tabular}
\end{center}



\section {License}
\copyright 2012 Virginia Bioinformatics Institute. All rights reserved.

Note: License may change at a future date.
\end{document}
