\documentclass[11pt]{article}

\usepackage[parfill]{parskip}
\usepackage{comment}

\setlength{\topmargin}{-.5in}
\setlength{\textheight}{9in}
\setlength{\oddsidemargin}{.125in}
\setlength{\textwidth}{6.25in}

\newcommand {\cmd}[1] {\begin{quote}#1\end{quote}}

\begin{document}
\title{Open Genomics Engine - Framework Documentation}
\author{Lee Baker and David Mittelman\\
Virginia Bioinformatics Institute \\
leecb@vt.edu, david.mittelman@vt.edu}
\maketitle

\section {Introduction}
The Open Genomics Engine (OpenGE) is a open software framework for manipulating and interpreting high-throughput sequencing data. This documentation describes core functions that are integrated into the official distribution of the framework. For a background on the project visit http://www.openge.org. OpenGE is available as binaries for Linux and MacOS or as source. The source distribution contains further instructions for compiling and installing the software. Future releases will include VMware and Amazon EC2 images.
\section {Using OpenGE}
OpenGE consists of a collection of operations that can be invoked on the command line using the following syntax:

\cmd{openge \textit{command [options]}}

where \textit{command} is one of the commands listed in the following table, and \textit{options} are various flags and parameters that may be supplied depending on the command:

\begin{center}
\begin{tabular}{lp{3.5in}}
\hline
Command&Description\\ \hline
convert&Convert a between file formats\\
count&Count reads in a file\\
dedup&Mark (or remove) duplicate reads\\
help&View detailed help for any OpenGE command\\
mergesort&Merge and sort multiple input files\\
stats&Generate statistics for one or more files\\
version&Print out the version number\\
view&View a filtered subset of a file\\
\end{tabular}
\end{center}

\subsection {General}
The following flags are common for all OpenGE commands. Additional flags and parameters may be available for individual commands.
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-i \textit{filename}&{-}{-}in \textit{filename}&Input filename. Input files can also be specified without this flag. Defaults to stein if no input is specified.\\
-o \textit{filename}&{-}{-}out \textit{filename}&Output filename. Defaults to stdout if omitted. \\
-v&{-}{-}verbose&Display additional information to stderr during operation (e.g. progress indicators and other informational messages). Optional.\\
-T&{-}{-}tmpdir&Specify a directory to store temporary files. (default /tmp) Optional.\\
-t&{-}{-}threads&Set the number of threads to be used for multithreaded operations. Optional.\\
-d&{-}{-}nothreads&Disable multithreading.\\
\end{tabular}
\end{center}

When input or output files are required, stdin or stdout may be used by simply omitting \textit{filename}. For example, the following command:

\cmd{openge mergesort}

with no input specified, will read a file from stdin, operate on the file, and write the results to stdout. This enables commands to be chained together. For instance:

\cmd{openge dedup in.sam $|$ openge count}

will count the number of lines in a SAM file. Console redirection can be used as well. The following two commands are equivalent in a bash shell:

\cmd{openge mergesort a.bam -o c.bam}
\cmd{openge mergesort $<$ a.bam $>$ c.bam}

\subsection {File formats}

OpenGE supports both BAM and SAM formats as inputs for all commands. All results are written in BAM format.

Most commands support multiple files as inputs; these files should have identical headers; corrupted data may result from processing multiple BAM files with different headers. A warning will be displayed if the headers in your input files are not identical.

\section {Commands}

\subsection {convert}
Converts between file formats.

Example:
\cmd{openge convert in.sam -o out.bam}
\subsection {count}
The count command returns the number of reads contained in one or more files. No parameters (other than input files) are required.

Example:
\cmd{openge count a.bam}
\cmd{openge count b.bam c.bam d.bam}

\subsection {dedup}
The dedup command either marks duplicate reads (compatible with Picard) or removes the duplicate reads. 

Duplicates are marked by setting the 'duplicate' bitflag for each read found to be a duplicate. If the -r or --remove flags are specified, the read is omitted from the output bamfile. The duplicate identification 

\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-r&{-}{-}remove&Remove duplicates instead of only marking them.\\
\end{tabular}
\end{center}

For example,
\cmd{openge dedup a.bam -o z.bam} 
identifies duplicates in a.bam, and writes all reads to z.bam with duplicates identified.

\subsection {help}
The help command shows available options for any OpenGE command. 

\cmd{openge help \textit{[command]}}

For example:

\cmd{openge help mergesort}

\cmd{openge help count}

Available commands can be seen in the table at the top of this section.


\subsection {mergesort}
This command generates a single file from one or more input files. The supplied input files are merged and then sorted by read position or region.

Parameters:
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-c \textit{level}&{-}{-}compression \textit{level}&Compression level- defaults to 6. Valid levels are 0-9, and correspond to zlib's deflate compression levels. \\
-r \textit{region}&{-}{-}region \textit{region}&Genomic region to use\\
-n \textit{reads}&{-}{-}n \textit{reads}&Number of reads to use per temporary file. Defaults to 500,000.\\
-C&{-}{-}compresstempfiles&Compress temporary files. By default, temporary files are not compressed.\\
-m&{-}{-}markduplicates&Mark duplicates after sorting.\\
-R&{-}{-}removeduplicates&Mark and remove duplicates after sorting.\\
\end{tabular}
\end{center}

The number of reads per temporary file can have a substantial effect on speed when processing large datasets. Increasing the number of reads per file produces larger (and fewer) temporary files at the expense of memory, but reduces the processing time.

\subsection{stats}
Prints out statistics about a file. No flags are required.
The stats command generates statistics for one or more input files. The computed statistics are summarized below.

Example:

\cmd{openge stats a.bam}
\cmd{openge stats a.bam b.bam c.bam {-}{-}inserts {-}{-}verbose}
Parameters:
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-I \textit{level}&{-}{-}inserts \textit{level}&Include detailed stats about inserts.\\
\end{tabular}
\end{center}

\subsection{version}
Prints out the version number and the date and time that this version was compiled.

\subsection{view}
The view command returns a portion of a bamfile optionally filtered by region, and then optionally limited to a smaller number of reads.

Parameters:
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-n \textit{number}&{-}{-}count\textit{number}&Number of reads to include in the generated file. Defaults to include the entire file.\\
-r\textit{region}&{-}{-}region \textit{region}&Region string (see below for format)\\
\end{tabular}
\end{center}

\subsubsection{Region string format}
Region strings are formatted exactly like the equivalent bamtools region strings, and this section is an excerpt from the bamtools documentation.

A proper REGION string can be formatted like any of the following examples (where 'chr1' is the name of a reference (not its ID)and '' is any valid integer position within that reference.):

\begin{center}
\begin{tabular}{lp{3.5in}}
{-}{-}region chr1&only alignments on (entire) reference 'chr1'\\
{-}{-}region chr1:500&only alignments overlapping the region starting at chr1:500 and continuing to the end of chr1\\
{-}{-}region chr1:500..1000&only alignments overlapping the region starting at chr1:500 and continuing to chr1:1000\\
{-}{-}region chr1:500..chr3:750&only alignments overlapping the region starting at chr1:500 and continuing to chr3:750.\\
\end{tabular}
\end{center}

This 'spanning' region assumes that the reference specified as the right boundary will occur somewhere in the file after the left boundary. On a sorted file, a region of 'chr4:500..chr2:1500' will produce undefined (incorrect) results. So don't do it. :)

\newpage
\section {License}
This software is made available through the Virginia Tech non-commerical license. See openge/license.txt for more details. 

\copyright 2012 Virginia Bioinformatics Institute. All rights reserved.
\end{document}
