\documentclass[11pt]{article}

\usepackage[parfill]{parskip}
\usepackage{comment}

\setlength{\topmargin}{-.5in}
\setlength{\textheight}{9in}
\setlength{\oddsidemargin}{.125in}
\setlength{\textwidth}{6.25in}

\newcommand {\cmd}[1] {\begin{quote}#1\end{quote}}

\begin{document}
\title{Open Genomics Engine - Framework Documentation}
\author{Lee Baker and David Mittelman\\
Virginia Bioinformatics Institute \\
leecb@vt.edu, david.mittelman@vt.edu}
\maketitle

\section {Introduction}
The Open Genomics Engine (OpenGE) is an open software framework for manipulating and interpreting high-throughput sequencing data. This documentation describes core functions that are integrated into the official distribution of the framework. OpenGE is available as binaries for Linux and MacOS or as source. The source distribution contains further instructions for compiling and installing the software. Future releases will include VMware and Amazon EC2 images.
\section {Using OpenGE}
OpenGE consists of a collection of operations that can be invoked on the command line using the following syntax:

\cmd{openge \textit{command [options]}}

where \textit{command} is one of the commands listed in the following table, and \textit{options} are various flags and parameters that may be supplied depending on the command:

\begin{center}
\begin{tabular}{lp{3.5in}}
\hline
Command&Description\\ \hline
count&Count reads in an alignment\\
coverage&Compute coverage statistics for an alignment\\
dedup&Mark (or remove) duplicate reads\\
help&View detailed help for any OpenGE command\\
mergesort&Merge and sort multiple input files\\
localrealign&Realign misaligned reads due to indels\\
stats&Generate statistics for one or more files\\
version&Print out the version number\\
view&View a filtered subset of a file\\
\end{tabular}
\end{center}

\subsection {General}
The following flags are common for all OpenGE commands. Additional flags and parameters may be available for individual commands.
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-i \textit{filename}&{-}{-}in \textit{filename}&Input filename. Input files can also be specified without this flag. Defaults to stein if no input is specified.\\
-o \textit{filename}&{-}{-}out \textit{filename}&Output filename. Defaults to stdout if omitted. \\
-v&{-}{-}verbose&Display additional information to stderr during operation (e.g. progress indicators and other informational messages). Optional.\\
-c \textit{level}&{-}{-}compression \textit{level}&Compression level- defaults to 6. Valid levels are 0-9, and correspond to zlib's deflate compression levels. Optional.\\
-T&{-}{-}tmpdir&Specify a directory to store temporary files. (default /tmp) Optional.\\
-t&{-}{-}threads&Set the number of threads to be used for multithreaded operations. Optional.\\
-d&{-}{-}nothreads&Disable multithreading. Optional.\\
&{-}{-}nosplit&Disable splitting by chromosome (see below). Optional.\\
-F \textit{format}&{-}{-}format \textit{format}&Select file format. Optional.\\
\end{tabular}
\end{center}

When running some commands, OpenGE may process chromosomes in separate threads in order to increase speed. Whether or not this occurs depends on the number of cores available on your machine. This increases the amount of memory consumed- you can disable this with the {-}{-}nosplit option.

When input or output files are required, stdin or stdout may be used by simply omitting \textit{filename}. For example, the following command:

\cmd{openge mergesort}

with no input specified, will read a file from stdin, operate on the file, and write the results to stdout. This enables commands to be chained together. For instance:

\cmd{openge dedup in.sam $|$ openge count}

will count the number of lines in a SAM file. Console redirection can be used as well. The following two commands are equivalent in a bash shell:

\cmd{openge mergesort a.bam -o c.bam}
\cmd{openge mergesort $<$ a.bam $>$ c.bam}

\subsection {File formats}

OpenGE supports both BAM and SAM formats as inputs for all commands. By default, results are written in BAM format; experimental SAM and FASTQ support is also present. To write in SAM format, specify a filename with a .sam or .SAM extension; for FASTQ format, use a .fastq or .FASTQ extension.

Most commands support multiple files as inputs; these files should have identical headers; corrupted data may result from processing multiple BAM or SAM files with different headers. A warning will be displayed if the headers in your input files are not identical.

Conversion between formats can be performed with the view command. For instance:
\cmd{openge view in.bam out.sam}
converts the file in.bam to SAM format.

\cmd{openge view in.bam out.fastq}
converts the file in.bam to FASTQ format.

\subsubsection {Selecting the output format}
OpenGE tries to automatically select the correct output format by guessing from the filename. Using extensions of .bam, .sam and .fastq (case insensitive) for the output filename will result in the appropriate file format being selected. Alternatively, you can specify a format with the {-}{-}format parameter. If there is neither a valid extension or a format specified on the command line, the output defaults to BAM (or SAM if using the view command and writing to stdout).

To summarize, the following methods are used to select file type (in order of preference):
\begin{itemize}
\item If a file format is specified on the command line with {-}{-}format, the specified format is used
\item If the file has an extension of .bam, .sam, or .fastq (case insensitive), that format is used
\item If the command is view, and data is being written to stdout, SAM format is used for output
\item Finally, BAM format is used for anything else
\end{itemize}

\section {Commands}

\subsection {count}
The count command returns the number of reads contained in one or more files. No parameters (other than input files) are required.

Example:
\cmd{openge count a.bam}
\cmd{openge count b.bam c.bam d.bam}

\subsection {coverage}
The coverage command reports the read depth as a function of reference genome location. 

Optionally, detailed information on the coverage of every base of the reference genome can be reported. This file can be very large and time consuming to generate, so it is not enabled by default. 

\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-o  \textit{filename}&{-}{-}output \textit{filename}&Output coverage data base-by-base to a TSV file.\\
&{-}{-}omituncoveredbases&Omit bases from the output file that are not covered by any read\\
-V&{-}{-}verifymapping&Verify that mapping was performed correctly (see below)\\
-S&{-}{-}strict&Use strict verification- only use first coordinate\\
-B \textit{size}&{-}{-}binsize \textit{size}&Bin output data into bins of \textit{size} bases\\
\end{tabular}
\end{center}

Coverage can also be used to verify that mappings have been performed correctly. To do this, reads must be named with a name that starts with the chromosome name and contains the position of the read start and end, each separated with underscores ('\_'). For example, "chr1\_2345\_2388" is valid, as is "chr99\_3350203\_3350299\_long\_name\_9993352:2334". A read is marked as correctly mapped if the mapped chromosome matches the chromosome in the name, and the position is within 5 bases of either of the positions identified in the name (without {-}{-}strict), or within five bases of the first position (with {-}{-}strict). The number of correctly mapped bases is included in the output file, if enabled (ie, if -o and -V are used).

To only report the coverage statistics for each chromosome:
\cmd{openge coverage a.bam}

Or, to output the coverage statistics for each base, and verify the mapping correctness:
\cmd{openge coverage b.bam -o bases.tsv -V}

\subsection {dedup}
The dedup command either marks duplicate reads (compatible with Picard) or removes the duplicate reads. 

Duplicates are marked by setting the 'duplicate' bitflag for each read found to be a duplicate. If the -r or --remove flags are specified, the read is omitted from the output bamfile. The duplicate identification 

\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-R&{-}{-}removeduplicates&Remove duplicates instead of only marking them.\\
\end{tabular}
\end{center}

For example,
\cmd{openge dedup a.bam -o z.bam} 
identifies duplicates in a.bam, and writes all reads to z.bam with duplicates identified.

\subsection {help}
The help command shows available options for any OpenGE command. 

\cmd{openge help \textit{[command]}}

For example:

\cmd{openge help mergesort}

\cmd{openge help count}

Available commands can be seen in the table at the top of this section.


\subsection {mergesort}
This command generates a single file from one or more input files. The supplied input files are merged and then sorted by read position or region.

Parameters:
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-r \textit{region}&{-}{-}region \textit{region}&Genomic region to use\\
-q \textit{min\_mapq}&{-}{-}mapq \textit{min\_mapq}&Minimum mapping quality for a read to be included in the mergesort.\\
-n \textit{reads}&{-}{-}n \textit{reads}&Number of reads to use per temporary file. Defaults to 500,000.\\
-C&{-}{-}compresstempfiles&Compress temporary files. By default, temporary files are not compressed.\\
-M&{-}{-}markduplicates&Mark duplicates after sorting.\\
-R&{-}{-}removeduplicates&Mark and remove duplicates after sorting.\\
\end{tabular}
\end{center}

The number of reads per temporary file can have a substantial effect on speed when processing large datasets. Increasing the number of reads per file produces larger (and fewer) temporary files at the expense of memory, but reduces the processing time.

\subsection {localrealign}
This command realigns reads around indels in order to minimize the number of mismatching bases. Since localrealignment uses multiple reads to form a consensus when realigning, it may produce more correct mappings than the original mapper.

This function is intended to be compatible with GATK's IndelRealigner tool.
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-R \textit{filename}&{-}{-}reference \textit{filename}&Reference genome in FASTA format. Required.\\
-L \textit{filename}&{-}{-}intervals \textit{filename}&Target intervals to realign. See below for format. Required.\\
\end{tabular}
\end{center}

For example, to realign a file a.bam (generating a.realigned.bam), use the following command:
\cmd{openge localrealign --reference dmel.fasta --intervals a.intervals a.bam -o a.realigned.bam}

\subsubsection {Intervals file format}
The intervals file is a text file listing intervals around which we should realign. Lines must be in the following format:
\textit{chr:\#} or \textit{chr:\#-\#} with no whitespace. This matches the format generated by GATK's RealignerTargetCreator.

\subsection{stats}
Prints out statistics about a file. No flags are required.
The stats command generates statistics for one or more input files. The computed statistics are summarized below.

Example:

\cmd{openge stats a.bam}
\cmd{openge stats a.bam b.bam c.bam {-}{-}inserts {-}{-}lengths {-}{-}verbose}
Parameters:
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-I&{-}{-}inserts&Include detailed stats about inserts.\\
-L&{-}{-}lengths&Include the distribution of read sizes.\\
\end{tabular}
\end{center}

For example:
\cmd {openge stats a.bam b.bam -I -L}
gives results similar to this:
\cmd {Total reads:            15419\\
Mapped reads:           14812 ( 96.1\%)\\
Forward strand:          7901 ( 51.2\%)\\
Reverse strand:          7518 ( 48.8\%)\\
Failed QC:                  0 (  0.0\%)\\
Duplicates:                 0 (  0.0\%)\\
Paired-end reads:       15419 (100.0\%)\\
'Proper-pairs':         13856 ( 89.9\%)\\
Both pairs mapped:      14205 ( 92.1\%)\\
Read 1:                  7697\\
Read 2:                  7722\\
Singletons:               607 (  3.9\%)\\
Read lengths:\\
    45bp:                2327 ( 15.1\%)\\
    64bp:                1699 ( 11.0\%)\\
    75bp:               11393 ( 73.9\%)\\
Insert size (absolute value):
    Mean:               218.7
    Median:             189.0}

\subsection{version}
Prints out the version number and the date and time that this version was compiled.

\subsection{view}
The view command returns a portion of a file optionally restricted to:
\begin{itemize}
\item Reads mapped to a specified chromosome, or region of a chromosome ({-}{-}region)
\item Reads with Mapq values above a threshold ({-}{-}mapq)
\item Reads of a certain length or range of lengths ({-}{-}length)
\item Limited to a certain number of results ({-}{-}count)
\end{itemize}
or any combination of the above.

The view command can also trim bases from the beginning or end of reads. If the amount being trimmed from the beginning and end combined exceeds the length of the read, the read will be omitted.

Parameters:
\begin{center}
\begin{tabular}{llp{3.5in}}
\hline
Flag&Long flag&Description\\ \hline
-n \textit{number}&{-}{-}count\textit{number}&Number of reads to include in the generated file. Defaults to include the entire file.\\
-r \textit{region}&{-}{-}region \textit{region}&Region string (see below for format)\\
-q \textit{min\_mapq}&{-}{-}mapq \textit{min\_mapq}&Minimum mapq allowed\\
-l \textit{length}&{-}{-}length \textit{length}&Allowed range of read lengths (see below for format)\\
-B \textit{length}&{-}{-}trimbegin \textit{length}&Trim \textit{length} bases from the beginning of all reads\\
-E \textit{length}&{-}{-}trimend \textit{length}&Trim \textit{length} bases from the end of all reads\\
\end{tabular}
\end{center}

Trimming with the {-}{-}trimbegin and {-}{-}trimend parameters is only supported for the FASTQ output format at this time.

\subsubsection{Region string format}
Region strings are formatted similarly to the equivalent bamtools region strings, and this section is an excerpt from the bamtools documentation.

A proper region string can be formatted like any of the following examples:

\begin{center}
\begin{tabular}{lp{3.5in}}
{-}{-}region chr1&only alignments on (entire) reference 'chr1'\\
{-}{-}region chr1:500&only alignments overlapping the region starting at chr1:500 and continuing to the end of chr1\\
{-}{-}region chr1:500..1000&only alignments overlapping the region starting at chr1:500 and continuing to chr1:1000\\
\end{tabular}
\end{center}

\subsubsection{Length format strings}
The following formats can be used to 
\begin{center}
\begin{tabular}{lp{3.5in}}
{-}{-}length 64&Only allow reads with length of 64\\
{-}{-}length +64&Only allow reads with length greater than 64\\
{-}{-}length -64&Only allow reads with length less than 64\\
{-}{-}length 64-72&Only allow reads with length between 64 and 72 (inclusive)\\
\end{tabular}
\end{center}

\section {License}
This software is made available through the Virginia Tech non-commerical license. See openge/license.txt for more details. 

\copyright 2012 Virginia Bioinformatics Institute. All rights reserved.
\end{document}
